workflow/executor/executor.go

```diff
@@ -780,17 +781,38 @@ func (we *WorkflowExecutor) CaptureScriptResult(ctx context.Context) error {
 	return nil
 }
 
-// reportOutputs updates the WorkflowTaskResult (or falls back to annotate the Pod)
-func (we *WorkflowExecutor) reportOutputs(ctx context.Context, logArtifacts []wfv1.Artifact) error {
+// FinalizeOutput adds a label or annotation to denote that outputs have completed reporting.
+func (we *WorkflowExecutor) FinalizeOutput(ctx context.Context) {
+	err := retryutil.OnError(wait.Backoff{
+		Duration: time.Second,
+		Factor:   2,
+		Jitter:   0.1,
+		Steps:    5,
+		Cap:      30 * time.Second,
+	}, errorsutil.IsTransientErr, func() error {
+		err := we.patchTaskResultLabels(ctx, map[string]string{
+			common.LabelKeyReportOutputsCompleted: "true",
+		})
+		if apierr.IsForbidden(err) {
+			log.WithError(err).Warn("failed to patch task set, falling back to legacy/insecure pod patch, see https://argoproj.github.io/argo-workflows/workflow-rbac/")
+			// Only added as a backup in case LabelKeyReportOutputsCompleted could not be set
+			err = we.AddAnnotation(ctx, common.AnnotationKeyReportOutputsCompleted, "true")
+		}
+		return err
+	})
+	if err != nil {
+		we.AddError(err)
+	}
+}
+
+// ReportOutputs updates the WorkflowTaskResult (or falls back to annotate the Pod)
+func (we *WorkflowExecutor) ReportOutputs(ctx context.Context, logArtifacts []wfv1.Artifact) error {
 	outputs := we.Template.Outputs.DeepCopy()
 	outputs.Artifacts = append(outputs.Artifacts, logArtifacts...)
 	return we.reportResult(ctx, wfv1.NodeResult{Outputs: outputs})
 }
 
 func (we *WorkflowExecutor) reportResult(ctx context.Context, result wfv1.NodeResult) error {
-	if !result.Outputs.HasOutputs() && !result.Progress.IsValid() {
-		return nil
```

Comment:
hmm, I wonder if removing this is check is okay or not. I know you wanted to be able to add the annotation, which is fine. But it does seem like if this were to resolve as true, we will do some of the things below which we weren't doing before. But maybe the check was just here to preemptively return in case there seemed to be nothing to do? (I'm not sure what causes Progress to be invalid.)

Reply:
My assumption was that this was here to return early if there was nothing to do. I removed this before I realized that a task result is created before `ReportOutputs` of the wait container (when reporting progress for the first time, which happens early on in the wait container execution). I'll add it back.

I guess progress can be invalid if there are no tasks. Not sure if N can ever be greater than M. I'm pretty sure the progress should always be set at this point (it happens early on in wait container with `monitorProgress`)

IsValid logic is
```
in != "" && in.N() >= 0 && in.N() <= in.M() && in.M() > 0
```
where  `Progress in N/M format. N is number of task complete. M is number of tasks.`

Reply:
Thanks for going through the logic. It does seem safe to me to remove this line. If you don't have Outputs and you don't have Tasks, then it seems like it would probably be fine to still call `upsertTaskResult()` below, right?

Reply:
Yeah, it would just create a shell of a `WorkflowTaskResult`